pkg/bootstrap/bootstrap_config.go

```diff
@@ -218,43 +218,104 @@ func StoreHostPort(host, port, field string, opts map[string]interface{}) {
 	opts[field] = fmt.Sprintf("{\"address\": \"%s\", \"port_value\": %s}", host, port)
 }
 
-type setMetaFunc func(m map[string]string, key string, val string)
+type setMetaFunc func(m map[string]interface{}, key string, val string)
 
-func extractMetadata(envs []string, prefix string, set setMetaFunc, meta map[string]string) {
+func extractMetadata(envs []string, prefix string, set setMetaFunc, meta map[string]interface{}) {
 	metaPrefixLen := len(prefix)
 	for _, env := range envs {
-		if strings.HasPrefix(env, prefix) {
-			v := env[metaPrefixLen:]
-			parts := strings.SplitN(v, "=", 2)
-			if len(parts) != 2 {
-				continue
-			}
-			metaKey, metaVal := parts[0], parts[1]
+		if !shouldExtract(env, prefix) {
+			continue
+		}
+		v := env[metaPrefixLen:]
+		if !isEnvVar(v) {
+			continue
+		}
+		metaKey, metaVal := parseEnvVar(v)
+		set(meta, metaKey, metaVal)
+	}
+}
+
+type istioMetadata struct {
+	CanonicalTelemetryService string            `json:"canonical_telemetry_service,omitempty"`
+	IP                        string            `json:"ip,omitempty"`
+	Labels                    map[string]string `json:"labels,omitempty"`
+	Name                      string            `json:"name,omitempty"`
+	Namespace                 string            `json:"namespace,omitempty"`
+	ServiceAccount            string            `json:"service_account,omitempty"`
+}
 
-			set(meta, metaKey, metaVal)
+func shouldExtract(envVar, prefix string) bool {
+	// this will allow transition from current method of exposition in the future
+	// Example:
+	// if strings.HasPrefix(envVar, "ISTIO_METAJSON_LABELS") {
+	// 	return false
+	// }
+	return strings.HasPrefix(envVar, prefix)
+}
+
+func isEnvVar(str string) bool {
+	return strings.Contains(str, "=")
+}
+
+func parseEnvVar(varStr string) (string, string) {
+	parts := strings.SplitN(varStr, "=", 2)
+	if len(parts) != 2 {
+		return varStr, ""
+	}
+	return parts[0], parts[1]
+}
+
+func jsonStringToMap(jsonStr string) (m map[string]string) {
+	err := json.Unmarshal([]byte(jsonStr), &m)
+	if err != nil {
+		log.Warnf("Env variable with value %q failed json unmarshal: %v", jsonStr, err)
+	}
+	return
+}
+
+func extractIstioMetadata(envVars []string) istioMetadata {
+	im := istioMetadata{}
+	for _, varStr := range envVars {
+		name, val := parseEnvVar(varStr)
+		switch name {
+		case "INSTANCE_IP":
+			im.IP = val
+		case "ISTIO_METAJSON_LABELS":
+			m := jsonStringToMap(val)
+			im.Labels = m
+			if svc, found := m["istioTelemetryService"]; found {
```

Comment:
nit: I don't think you need to check for `found` ... defaults to an empty string if not found.

Reply:
fair enough. updated.